core/src/main/java/hudson/model/UpdateCenter.java

```diff
@@ -1502,23 +1502,38 @@ public Plugin getPlugin() {
         @Override
         public void run() {
             try {
-                plugin.getInstalled().enable();
-            } catch (IOException e) {
-                LOGGER.log(Level.SEVERE, "Failed to enable " + plugin.getDisplayName(), e);
-                error = e;
-            }
-            
-            if (dynamicLoad) {
-                try {
-                    // remove the existing, disabled inactive plugin to force a new one to load
-                    pm.dynamicLoad(getDestination(), true);
-                } catch (Exception e) {
-                    LOGGER.log(Level.SEVERE, "Failed to dynamically load " + plugin.getDisplayName(), e);
-                    error = e;
-                    requiresRestart = true;
+                PluginWrapper installed = plugin.getInstalled();
+                synchronized(installed) {
+                    if (!installed.isEnabled()) {
+                        try {
+                            installed.enable();
+                        } catch (IOException e) {
+                            LOGGER.log(Level.SEVERE, "Failed to enable " + plugin.getDisplayName(), e);
+                            error = e;
+                            status = new Failure(e);
+                        }
+                        
+                        if (dynamicLoad) {
+                            try {
+                                // remove the existing, disabled inactive plugin to force a new one to load
+                                pm.dynamicLoad(getDestination(), true);
+                            } catch (Exception e) {
+                                LOGGER.log(Level.SEVERE, "Failed to dynamically load " + plugin.getDisplayName(), e);
+                                error = e;
+                                requiresRestart = true;
+                                status = new Failure(e);
+                            }
+                        } else {
+                            requiresRestart = true;
+                        }
+                    }
                 }
-            } else {
+                status = new Success();
```

Comment:
Yes, this was an epic fail. I had `if (status instanceof Pending)` and idiotically removed it on a late Friday.